src/polyswarmd/websockets/messages.py

```diff
@@ -1,386 +1,493 @@
-try:
-    import ujson as json
-except ImportError:
-    import json
-from typing import Any
-
-from polyswarmartifact import ArtifactType
-from polyswarmartifact.schema import Bounty as BountyMetadata, Assertion as AssertionMetadata
-
-from .json_schema import copy_with_schema
+from typing import (
+    Any,
+    Callable,
+    ClassVar,
+    Dict,
+    Generic,
+    Mapping,
+    Optional,
+    cast,
+)
+import ujson
 
+from requests_futures.sessions import FuturesSession
 
-class WebsocketMessage:
+from polyswarmartifact import ArtifactType
+from polyswarmartifact.schema import Assertion as AssertionMetadata
+from polyswarmartifact.schema import Bounty as BountyMetadata
+
+from .json_schema import PSJSONSchema, SchemaDef
+from .message_types import (
+    ClosedAgreementMessageData,
+    D,
+    E,
+    EventData,
+    FeesUpdatedMessageData,
+    InitializedChannelMessageData,
+    LatestEventMessageData,
+    NewAssertionMessageData,
+    NewBountyMessageData,
+    NewDepositMessageData,
+    NewVoteMessageData,
+    NewWithdrawalMessageData,
+    QuorumReachedMessageData,
+    RevealedAssertionMessageData,
+    SettledBountyMessageData,
+    SettleStateChallengedMessageData,
+    StartedSettleMessageData,
+    TransferMessageData,
+    WebsocketEventMessage,
+    WindowsUpdatedMessageData,
+)
+
+
+class WebsocketMessage(Generic[D]):
     "Represent a message that can be handled by polyswarm-client"
     # This is the identifier used when building a websocket event identifier.
-    _ws_event: str
-    __slots__ = ('data')
+    event: ClassVar[str]
+    __slots__ = {'message': bytes}
```

Comment:
I am not a huge fan of `__slots__` unless you want to enforce the fact that you can't create dynamic attributes. Is that the case here? If only for performance, I would not add this as it changes some of the internal behavior of the class, creates a larger burden of maintenance (two places to edit when adding properties to the class) and does not mingle well with inheritance. I am not opposed to the use, but I have seen plenty of code that relies on `__dict__` and this will not play along nicely, so I would rather keep a more predictable behavior than extract 15% performance on attribute access. You can dismiss this comment if you feel like the use is justifiable.

Reply:
The idea behind `__slots__` here is that we'll be creating a few of these objects. I haven't ran the numbers but documentation surrounding `__slots__` has some remarkable claims about memory-savings.

The inheritance bit I've already run into - what else can I do to reduce the amount of space used by an object instance?